### ISSUE TITLE
DASD NED 1 Serial Number Problem

### ISSUE THREAD

**Issue by Peter-J-Jansen** [SDL-Hercules-390/hyperion#243]
Commit c5a61c8 as a result of Issue #166 (DASD File Protect Errors) introduced a randomly generated serial number in NED 1 available through the RCD CCW command.  When that very same DASD is used via Hercules' Shared DASD support by other z/VM SSI nodes, a serial number of zero is currently the result. This makes such other z/VM SSI nodes believe this is a different DASD, and such nodes do not complete their IPL process; such nodes get into an "Isolated" status.

Different solutions to this issue may be possible, but the end result must be that the NED 1 for any DASD must yield identical results for any shared usage of that DASD.

**Comment by Peter-J-Jansen** [comment]
As the current DASD serial number is randomly generated by Hercules, the Shared DASD server could be amended with providing the configuration data for the RCD CCW command to all Shared DASD users. This is probably not a trivial change, and is it really needed ? If not, then the proposal would be to just set the serial number field to 12 zeros ('0'). (And that would be a trivial change.)

Commit c5a61c8 also corrected other NED bit settings, which this author now believes is what fixes the problem as reported in Issue #166. (This should be verified.) This is based on the assumption that as long as each DASD has a unique NED 1, and all shared users of the same DASD receive identical NED 1 via RCD, the DASD serial number in NED 1 does not matter. The uniqueness of the DASD NED 1 is guaranteed by the presence of the unique CCUU address of the DASD which is also in the NED 1.

The author has tested this "zero DASD serial number" approach in a 3-node SSI z/VM cluster, with each z/VM node successfully running a base sysplex z/OS 2.3 system 2nd level.

Furthermore, additional CTCE device improvements will be committed next month. This now also include support for Ficon CTC adapters (FCTC), for which z/OS requires RCD support with their own NED requirements (please see SB10-7034-05 ESCON and FICON Channel-to-Channel Reference).  Here too, a zero serial number for the FCTC devices works flawlessly for the same reason as in my above assumption.

Hence my proposal for this "zero DASD serial number" solution.

Cheers,

Peter

**Comment by fbi-ranger** [comment]
In z/PDT the EQID is created by the number of the [manager]statement, a sort of grouping of devices to a CU, counted and the under that manager statement the position of the device put together to produce a unique identifier of a device. 

For Hercules it could be a counter for each device class. Such as DASD or OSA since there are NO CU controls in the config file. Then the EQID is dependen on its position in the config file. 
Or you may introduce an optional parameter for the device to specify a specific  EQID. 

For a SSI cluster this EQIDs are essential.

**Comment by Peter-J-Jansen** [comment]
So it appears that z/PDT uses their equivalent of the Hercules configuration file to derive input for z/VM's EQID's, and not something stored in the Linux files emulating the DASD device.

The DASD EQID's in z/VM are by default CP generated, and appear to be derived from the NED 1 data as returned by a Read Configuration Data (RCD) CCW command. (z/VM also supports manually specified 8-character EQID's in the RDEVICE configuration statement, a feature which could be used to avoid the current Hercules different NED 1 data for DASD's emulated by a file vs. Shared DASD's.)

For ease of reference, I repeat here the z/VM errors I encountered were when IPL-ing the 2nd and 3rd SSI members using Hercules Shared DASD, after the 1st member was IPL'd using the DASD file (by the same Hercules also functioning as the Shared DASD server) : 

```
19:04:34 HCPXDI6641E This system contains EQID 00339000AHRCZZ000000000000000016350000000000000D0A, volser M01S01 in slot 10 and the SSI cluster contains EQID 00339000AHRCZZ000036849735405516350000000000000D0A, volser M01S01 in this slot.
19:04:34 HCPXDI6641E This system contains EQID 00339000AHRCZZ000000000000000026350000000000000D0A, volser M02S01 in slot 11 and the SSI cluster contains EQID 00339000AHRCZZ000023232546720726350000000000000D0A, volser M02S01 in this slot.
19:04:34 HCPXDI6641E This system contains EQID 00339000AHRCZZ000000000000000036350000000000000D0A, volser M03S01 in slot 12 and the SSI cluster contains EQID 00339000AHRCZZ000056543566442336350000000000000D0A, volser M03S01 in this slot.
19:04:34 HCPPLM1663E SSI function JOIN has failed for service XDISK.
19:04:34 HCPPLM1697I The state of SSI system V3 has changed from JOINING to ISOLATED
```
Hercules is reported to currently generate a hopefully unique serial number at DASDINIT time. My DASD's files predate that feature, in which case the serial number which ends up in the NED 1 data is generated at DASD open time, e.g. as shown in the above "...368497354055..." for M01S01.  Apparantly, the very same DASD, when accessed as a Shared DASD by the 2nd and 3rd SSI nodes, does not follow this process, as it provides "...000000000000..." instead. Hence the error.

A simple, trivial test of mine, patching "dasdtab.c" as follows, solved the problem completely :

```
hercules@ubuntu:~$ diff -u ~/Qsync/Source/Originals/SDL-hyperion/dasdtab.c ~/Qsync/Source/Changes/SDL-hyperion/dasdtab.c
--- /home/hercules/Qsync/Source/Originals/SDL-hyperion/dasdtab.c	2019-08-25 12:23:39.949834900 +0200
+++ /home/hercules/Qsync/Source/Changes/SDL-hyperion/dasdtab.c	2019-08-28 21:48:37.979038900 +0200
@@ -509,7 +509,8 @@
     store_fw (buf, 0xdc010100);
     sprintf ((char *)&buf[4], "  %4.4X0%2.2XHRCZZ",
                         dev->ckdtab->devt, dev->ckdtab->model);
-    memcpy( &buf[18], dev->serial, 12 );
+/*  memcpy( &buf[18], dev->serial, 12 );      => Shared DASD difference !    PJJ */
+    memcpy( &buf[18], "000000000000", 12 );   /* Shared DASD must be equal   PJJ */
     for (i = 4; i < 30; i++)
         buf[i] = host_to_guest( isalnum( buf[i] ) ? buf[i] : '0' );
     store_hw(buf + 30, dev->devnum);        /* Uniquely tag within system */
hercules@ubuntu:~$
```
With this patch, the 3-member SSI z/VM cluster IPL'd perfectly.  I also tested z/OS 2.3 under those SSI z/VM nodes, by successfully IPL-ing 2 base sysplex nodes, one each in a SSI nodes. Their VM's had their z/OS ATTACHed, the first sysplex member had the z/OS DASD files in the Hercules configuration, the 2nd sysplex member had the same shared z/OS DASD as Hercules Shared DASD.  All works perfectly well and stable.

This, coupled with other RCD - NED experience (from FCTC devices as mentionned earlier), leads me to believe that the serial number field in the NED 1 data is not really required. The z/VM CP generated EQID's in the HCPXDI6441E error messages show that not only the serial number part of the NED 1 data is used, but also the DASD CCUU addresses.  Hence my proposal for the simple, trivial "zero DASD serial number" solution. However, this should deserve further testing, as I mentionned earlier.

The beauty of this "zero DASD serial number" solution is that : 

- is very trivial and simple

- introduces no new Hercules configuration stuff that one would need to take care of (which is what a z/PDT-like approach would result into)

- introduces no new code to the Shared DASD implementation (where so far I could not find out where the RCD command implementation should be added)

- hopefully still corrects the problem as reported in issue #166.  

Cheers,

Peter

**Comment by fbi-ranger** [comment]
Peter,

I think that the hardware should be emulated as close as possible to the real iron. Therefore the RCD should work identically on each instance of Hercules. Otherwise we might get further incompatibilities with new OS versions. 

I don't know exactly how the shared  DASD implementation is working but isn't it sending the CCWs to the DASD server instance? So also the read subsystem CCWs, where RDC and RCD information are send to the OS,  are transmitted to the DASD server, aren't they?

**Comment by Peter-J-Jansen** [comment]
I completely agree with that principle Florian. And Fish's serial number introduction for DASD, as a result of Issue #166, tried to do exactly that, albeit with a hopefully unique random generated serial number at dasdinit time, stored in the file emulating the DASD. For older DASD emulating files, at open time, the (again hopefully unique random generated) serial number is generated on the fly.

But up until now, looking at the code, I cannot explain why the latter process yields a zero serial number for Shared DASD. I do not think that a RCD CCW is not sent to the Shared DASD server. Also the Hercules "qd" command shows the zero serial number for Shared DASD, and a non-zero one for the DASD from the emulated DASD file. The "qd" command thus confirms the z/VM error message HCPXDI6641E contents that I showed earlier. I need to study the current code some more. Help will be very welcome here !

I presume that todays real iron DASD are always emulated, and I also presume that some of these emulations also have NED 1 / serial number problems, which might be the reason that the z/VM RDEFINE configuration statement allows the manual specification of the EQID for DASD, in which case the NED 1 info and its serial number is probably not used, at least not for the CP generated EQID's.  But I admit that that is a wild assumption of mine.

Cheers,

Peter

**Comment by Fish-Git** [comment]
@Peter-J-Jansen 
@fbi-ranger 

**FYI:** I have a fix that I'm working on which I hope to commit sometime this weekend that will completely resolve this issue.

Fixing the Shared DASD Server for RCD was actually quite trivial believe it or not:

```diff
--- hyperion-1/msgenu.h	2019-08-29 17:20:23.619375900 -0700
+++ hyperion-0/msgenu.h	2019-08-30 10:09:08.612863700 -0700
@@ -626,7 +626,7 @@
 #define HHC00713 "%1d:%04X Shared: error during channel program start"
 #define HHC00714 "%1d:%04X Shared: error during channel program end"
 #define HHC00715 "%1d:%04X Shared: remote error reading track %d"
-//efine HHC00716 (available)
+#define HHC00716 "%1d:%04X Shared: error retrieving serial number"
 #define HHC00717 "%1d:%04X Shared: error retrieving usage information"
 #define HHC00718 "%1d:%04X Shared: error writing track %d"
 #define HHC00719 "%1d:%04X Shared: remote error writing track %d %2.2X-%2.2X"
--- hyperion-1/shared.c	2019-08-29 17:20:23.712976000 -0700
+++ hyperion-0/shared.c	2019-08-29 19:15:38.000000000 -0700
@@ -275,6 +275,18 @@
     }
     dev->numdevid = rc;
 
+    /* Get the serial number */
+    rc = clientRequest (dev, dev->serial, sizeof(dev->serial),
+                        SHRD_QUERY, SHRD_SERIAL, NULL, NULL);
+    if (rc < 0)
+        goto init_retry;
+    else if (rc == 0 || rc > (int)sizeof(dev->serial))
+    {
+        // "%1d:%04X Shared: error retrieving serial number"
+        WRMSG( HHC00716, "S", LCSS_DEVNUM );
+        return -1;
+    }
+
     /* Indicate no active track */
     dev->cache = dev->bufcur = -1;
     dev->buf = NULL;
@@ -541,6 +553,18 @@
     }
     dev->numdevchar = rc;
 
+    /* Get the serial number */
+    rc = clientRequest (dev, dev->serial, sizeof(dev->serial),
+                        SHRD_QUERY, SHRD_SERIAL, NULL, NULL);
+    if (rc < 0)
+        goto init_retry;
+    else if (rc == 0 || rc > (int)sizeof(dev->serial))
+    {
+        // "%1d:%04X Shared: error retrieving serial number"
+        WRMSG( HHC00716, "S", LCSS_DEVNUM );
+        return -1;
+    }
+
     /* Indicate no active track */
     dev->cache = dev->bufcur = -1;
     dev->buf = NULL;
@@ -1958,6 +1982,11 @@
             serverSend (dev, ix, hdr, dev->devid, dev->numdevid);
             break;
 
+        case SHRD_SERIAL:
+            SHRD_SET_HDR (hdr, 0, 0, dev->devnum, id, sizeof( dev->serial ));
+            serverSend (dev, ix, hdr, dev->serial, (U32)sizeof( dev->serial ));
+            break;
+
         case SHRD_CKDCYLS:
             store_fw (buf, dev->ckdcyls);
             SHRD_SET_HDR (hdr, 0, 0, dev->devnum, id, 4);
--- hyperion-1/shared.h	2019-08-29 17:20:23.806576200 -0700
+++ hyperion-0/shared.h	2019-08-29 19:15:38.000000000 -0700
@@ -325,7 +325,7 @@
 /*-------------------------------------------------------------------*/
 
 #define SHARED_VERSION              0   /* Version level  (0 .. 15)  */
-#define SHARED_RELEASE              1   /* Release level  (0 .. 15)  */
+#define SHARED_RELEASE              2   /* Release level  (0 .. 15)  */
 
 /* Constraints                                                       */
 #define SHARED_DEFAULT_PORT      3990   /* Default shared port       */
@@ -383,6 +383,7 @@
 #define SHRD_DEVCHAR             0x41   /* Device characteristics    */
 #define SHRD_DEVID               0x42   /* Device identifier         */
 #define SHRD_USED                0x43   /* Device usage              */
+#define SHRD_SERIAL              0x44   /* Device serial number      */
 #define SHRD_CKDCYLS             0x48   /* CKD number cylinders      */
 #define SHRD_FBAORIGIN           0x4c   /* FBA origin                */
 #define SHRD_FBANUMBLK           0x4d   /* FBA number blocks         */
```

_**PLEASE NOTE**_ that while the above patch is only a _part_ of the overall fix, it is basically the _essential_ part that fixes the core of the problem. With only the above applied, the problem is essentially resolved. I encourage you to try it, Peter, to verify that it does indeed resolve your immediate problem. (I tried it myself and with the patch applied, a `qd` on either system displays the same serial number in the device NED.)

As always however, the devil is in the details. The complete "fix" involves much more, such as a new utility called `dasdser` to change the serial number for an existing dasd image (for those images which don't have a serial number or for when you simply want to assign a different value), a new optional `ser=nnnnnnnnnnnn` device statement option (for essentially temporarily accomplishing the same thing but mostly for uncompressed FBA/FBA64 dasds which don't have device headers (which is where the serial number is recorded)), documentation updates and of course the usual spin-off mostly cosmetic changes and tweaks I'm famous for.

As I said, I hope to be committing it sometime this weekend, but for now, please try the above patch because _**that**_ is what actually fixes the root of the problem.

Thanks.

**Comment by Peter-J-Jansen** [comment]
Dear Fish ,

Thank you ever so much for rescuing me !
After undoing my simple zero serial number fix, verifying I could reproduce the error, I then applied your patch, rebuild SDL Hercules on my 3 platforms (Windows 10 1903 VS2019, Ubuntu 18.04 and MacOS 10.13), and successfully IPL'd my 3-node SSI z/VM 6.3 on them. The Hercules qd command already showed it would work prior to the IPL. To test some more, on two of these z/VM nodes I then IPL'd z/OS 2.3 base sysplex members. The first one accesses it's ATTACHed DASD as Hercules files, the 2nd one as ATTACHed Hercules Shared DASD. Here too, no problems whatsoever.

Thanks again Fish !! Please feel free to commit your fix.

I know that you intend further enhancements to allow the user to set the serial number on DASD's, and that is fine of course. I can live perfectly well with the generated random serial numbers on the fly (as I now have them), also such serial numbers permanently stored at dasdinit time are perfectly OK with me. 

Cheers,

Peter

**Comment by Fish-Git** [comment]
Fixed by commit df0f6d26725236bcd9ef254a2d11b48232c1d3b3.

Closing.
